pkg/reconciler/pipelinerun/resources/validate_params.go

```diff
@@ -89,109 +88,60 @@ func ValidateObjectParamRequiredKeys(pipelineParameters []v1beta1.ParamSpec, pip
 	return nil
 }
 
-// ValidateParamArrayIndex validate if the array indexing param reference target is existent
-func ValidateParamArrayIndex(ctx context.Context, p *v1beta1.PipelineSpec, pr *v1beta1.PipelineRun) error {
+// ValidateParamArrayIndex validates if the param reference to an array param is out of bound.
+// error is returned when the array indexing reference is out of bound of the array param
+// e.g. if a param reference of $(params.array-param[2]) and the array param is of length 2.
+func ValidateParamArrayIndex(ctx context.Context, pipelineSpec *v1beta1.PipelineSpec, prParams []v1beta1.Param) error {
 	if !config.CheckAlphaOrBetaAPIFields(ctx) {
 		return nil
 	}
 
-	arrayParams := extractParamIndexes(p.Params, pr.Spec.Params)
+	// Collect all array params lengths
+	arrayParamsLengths := paramvalidation.ExtractParamArrayLengths(pipelineSpec.Params, prParams)
 
-	outofBoundParams := sets.String{}
-
-	// collect all the references
-	for i := range p.Tasks {
-		findInvalidParamArrayReferences(p.Tasks[i].Params, arrayParams, &outofBoundParams)
-		if p.Tasks[i].IsMatrixed() {
-			findInvalidParamArrayReferences(p.Tasks[i].Matrix.Params, arrayParams, &outofBoundParams)
+	paramsRefs := []string{}
+	for i := range pipelineSpec.Tasks {
+		paramsRefs = append(paramsRefs, extractParamValuesFromParams(pipelineSpec.Tasks[i].Params)...)
```

Comment:
Do we need to maintain the name of the task along with the name of the parameter here? There could be same param names `param1`,  one from `task1` and one from `task2`

Reply:
I don't think so. We're validating all the param refs to Pipelinerun params. And the task param can also have those refs in their values. And those refs have nothing to do with task name, or task param names. They only refer to pipeline's params. 
e.g.
```
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: deploy
spec:
  params:
    - name: pipelineParam1
      type: array
    - name: pipelineParam2
      type: array
  tasks:
    - name: task1
      params:
        - name: taskParam1
          value: '$(params.pipelineParam1[10])'
      taskSpec:
        params:
          - name: taskParam1
            type: string
    - name: task2
      params:
        - name: taskParam2
          value: '$(params.pipelineParam2[20])'
      taskSpec:
        params:
          - name: taskParam2
            type: string
```